### ISSUE TITLE
Suggest replacing `Option::and_then` with a literal `Some` with `Option::map` 

### ISSUE THREAD

**Issue by shepmaster** [rust-lang/rust-clippy#1459]
In code like

```rust
fn main() {
    let a = Some(21);
    let b = a.and_then(|a| Some(a + 21));
    println!("{:?}", b);
}
```

It should suggest changing to 

```rust
let b = a.map(|a| a + 21);
```

**Comment by drystone** [comment]
I've just found a load of these.

Same applies to `Result`
```
fn main()
    let a = Err("error".to_owned());
    a.or_else(|e| Err(format!("{}: {}, "this", e)));
    println!("{:?}", e);
}
```
Could suggest `map_err(|e|, format!("{}: {}", "this", e))`. And `Result::and_then` with a single `Ok()` branch could suggest `Result::map`

**Comment by smklein** [comment]
The "OPTION_AND_THEN_SOME" lint looks like it was added since this issue was opened, which I believe resolves this case.

**Comment by smklein** [comment]
https://github.com/rust-lang/rust-clippy/commit/7065239da55420e26adb7cb647fc7eb4e9c8798e

**Comment by shepmaster** [comment]
Clippy 0.0.212 (2020-04-18 891e1a8) says:

```
warning: using `Option.and_then(|x| Some(y))`, which is more succinctly expressed as `map(|x| y)`
 --> src/main.rs:3:13
  |
3 |     let b = a.and_then(|a| Some(a + 21));
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `a.map(|a| a + 21)`
  |
  = note: `#[warn(clippy::option_and_then_some)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#option_and_then_some
```

**Comment by shepmaster** [comment]
It does **not** report anything for the `Result` equivalent:

```rust
fn main() {
    let a = Err::<(), _>("error");
    let b = a.or_else(|_| Err("another error"));
    println!("{:?}", b);
}
```

Should that be moved to a new issue?
